src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+	char	   *raw_buf;
+	int			raw_buf_index;	/* next byte to process */
+	int			raw_buf_len;	/* total # of bytes stored */
+} CopyStateData;
+
+/* TODOL should be moved to copy.c */
+static StringInfo CopyGetLineBuf(CopyStateData *cs)
+{
+	return &cs->line_buf;
+}
+
+#define MAX_SHARDS 1001
+#define MAX_REPLICATION_FACTOR 8 /* should be less or equal than 32 */
+#define MAX_STMT_LEN 1024
+
+typedef struct {
+	int64 shardId;
+	int nReplicas;
+	int preparedMask;
+	PGconn* conn[MAX_REPLICATION_FACTOR];
+} CopyConnection;
+
+static uint32 shard_id_hash_fn(const void *key, Size keysize)
+{
+	return *(int32*)key ^ *((int32*)key+1);
+}
+
+/* 
+ * Construct hash table used for shardId->Connection mapping.
+ * We can not use connection cache from connection.c used by GteConnection because
+ * we need to establish multiple connections with each nodes: one connection per shard
+ */
+static HTAB *
+CreateShard2ConnectionHash()
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(int64);
+	info.entrysize = sizeof(CopyConnection);
+	info.hash = shard_id_hash_fn;
+
+	return hash_create("shard2conn", MAX_SHARDS, &info, HASH_ELEM | HASH_FUNCTION);
+}
+
+/*
+ * Execute statement with specified parameters and check its result
+ */
+static bool PgShardExecute(PGconn* conn, ExecStatusType expectedResult, char const* sql, ...)
+{
+	PGresult *result;
+	bool ret = true;
+	va_list args;
+	char buf[MAX_STMT_LEN];
+
+	va_start(args, sql);
+	vsnprintf(buf, sizeof(buf), sql, args);
+	result = PQexec(conn, buf);
+
+	if (PQresultStatus(result) != expectedResult)
+	{
+		ReportRemoteError(conn, result);
+		ret = false;
+	}
+	return ret;
+}
+
+/*
+ * Handle copy to/from distributed table
+ */
+bool PgShardCopy(CopyStmt *copyStatement, char const* query)
+{		 
+	RangeVar *relation = copyStatement->relation;
+	ListCell *shardIntervalCell;
+	ListCell *taskPlacementCell;
+	List *shardIntervalList;
+	int relationSuffixPos;
+	char *relationName;
+	char const *lowerQuery, *relationOcc;
+	PgCopyTransactionManager* tmgr = &PgCopyTmgrImpl[PgCopyTMGR];
+
+	if (relation != NULL)
+	{
+		bool failOK = true;
+		Oid tableId = RangeVarGetRelid(relation, NoLock, failOK);
+		bool isDistributedTable = false;
+
+		isDistributedTable = IsDistributedTable(tableId);
+		if (isDistributedTable)
+		{
+			HTAB* shard2conn;
+			MemoryContext tupleContext;
+			CopyState copyState;
+			bool nextRowFound = true;
+			TupleDesc tupleDescriptor = NULL;
+			uint32 columnCount = 0;
+			Datum *columnValues = NULL;
+			bool *columnNulls = NULL;
+			Var* partitionColumn;
+			Oid columnOid;
+			OpExpr *equalityExpr;
+			HASH_SEQ_STATUS hashCursor;
+			List *whereClauseList;
+			List *prunedList;
+			Node *rightOp;
+			Const *rightConst;
+			int64 failedShard = -1;
+			Relation rel;
+			StringInfo lineBuf;
+			CopyConnection* copyConn;
+			int i;
+
+			relationName = get_rel_name(tableId);
+			lowerQuery = lowerstr(query);
+			relationOcc =  strstr(lowerQuery, lowerstr(relationName));
```

Comment:
Yehhh...
Can't say that I almost existed by this code. Especially after replacing file name with STDIN. 
But reconstruction query string from options seems to be also non trivial task requiring much more code. And what is worser - if syntax of COPY command will be extended, this code also has to be changed...